src/formats/prometheus_text_format.erl

```diff
@@ -84,51 +114,18 @@ escape_label_value(LValue) when is_list(LValue) ->
 escape_label_value(Value) ->
     erlang:error({invalid_value, Value}).
 
-registry_collect_callback(Fd, Registry, Collector) ->
-    Callback = fun(MF) ->
-        emit_mf_prologue(Fd, MF),
-        emit_mf_metrics(Fd, MF)
-    end,
-    prometheus_collector:collect_mf(Registry, Collector, Callback).
-
-?DOC(false).
--spec emit_mf_prologue(Fd :: file:fd(), prometheus_model:'MetricFamily'()) -> ok.
-emit_mf_prologue(Fd, #'MetricFamily'{name = Name, help = Help, type = Type}) ->
-    Bytes = [
-        "# TYPE ",
-        Name,
-        " ",
-        string_type(Type),
-        "\n# HELP ",
-        Name,
-        " ",
-        escape_metric_help(Help),
-        "\n"
-    ],
-    file:write(Fd, Bytes).
-
-?DOC(false).
--spec emit_mf_metrics(file:fd(), prometheus_model:'MetricFamily'()) -> ok | {error, term()}.
-emit_mf_metrics(Fd, #'MetricFamily'{name = Name, metric = Metrics}) ->
-    %% file:write/2 is an expensive operation, as it goes through a port driver.
-    %% Instead a large chunk of bytes is being collected here, in a
-    %% way that triggers binary append optimization in ERTS.
-    Bytes = lists:foldl(
-        fun(Metric, Blob) ->
-            <<Blob/binary, (render_metric(Name, Metric))/binary>>
-        end,
-        <<>>,
-        Metrics
-    ),
-    file:write(Fd, Bytes).
+render_metrics(_Name, [], Bytes) ->
+    Bytes;
+render_metrics(Name, [Metric | Rest], Bytes) ->
+    render_metrics(Name, Rest, render_metric(Bytes, Name, Metric)).
```

Comment:
I like how this becomes tail-recursive. Note nevertheless, that if we want to be really mad about performance, the shuffling in the params ordering here is not ideal. The BEAM assigns each function input to X registers, so before calling `render_metric(Bytes, Name, Metric)`, it would have to shuffle X0 to X1, and X2 to X0. The JIT will turn this into CPU registers so it is insanely fast, but it will not reorder them into a faster version because it would mess up with tracing. So instead we could be calling here `render_metric(Name, Metric, Bytes)` and the only argument arrangement would be by extracting `Metric` into X1. Likely to be in the order of 3 CPU cycles the whole thing, but `gen_server` code takes great care to do this right so, why not 😄